modin/experimental/engines/omnisci_on_ray/frame/data.py

```diff
@@ -673,6 +689,83 @@ def _union_all(
 
         return new_frame
 
+    def _join_by_index(self, other_modin_frames, how, sort, ignore_index):
+        if how == "outer":
+            raise NotImplementedError("outer join is not supported in OmniSci engine")
+
+        lhs = self._maybe_materialize_rowid()
+        for rhs in other_modin_frames:
+            rhs = rhs._maybe_materialize_rowid()
+            if len(lhs._index_cols) != len(rhs._index_cols):
+                raise NotImplementedError(
+                    "join by indexes with different sizes is not supported"
+                )
+
+            if len(lhs._index_cols) == 1:
+                condition = lhs.ref(lhs._index_cols[0]).eq(rhs.ref(rhs._index_cols[0]))
+            else:
+                condition = OpExpr(
+                    "AND",
+                    [
+                        lhs.ref(lhs_col).eq(rhs.ref(rhs_col))
+                        for lhs_col, rhs_col in zip(lhs._index_cols, rhs._index_cols)
+                    ],
+                    get_dtype(bool),
+                )
+
+            exprs = lhs._index_exprs()
+            new_columns = []
+            for col in lhs.columns:
+                exprs[col] = lhs.ref(col)
+                new_columns.append(col)
+            for col in rhs.columns:
+                # Handle duplicating column names here. When user specifies
+                # suffixes to make a join, actual renaming is done in front-end.
+                new_col_name = col
+                rename_idx = 0
+                while new_col_name in exprs:
+                    new_col_name = f"{col}{rename_idx}"
+                    rename_idx += 1
+                exprs[new_col_name] = rhs.ref(col)
+                new_columns.append(new_col_name)
+
+            op = JoinNode(
+                lhs,
+                rhs,
+                how=how,
+                exprs=exprs,
+                condition=condition,
+            )
+
+            new_columns = Index.__new__(
+                Index, data=new_columns, dtype=self.columns.dtype
+            )
+            lhs = lhs.__constructor__(
+                dtypes=lhs._dtypes_for_exprs(exprs),
+                columns=new_columns,
+                index_cols=lhs._index_cols,
+                op=op,
+                force_execution_mode=self._force_execution_mode,
+            )
+
+        if sort:
+            lhs = lhs.sort_rows(
+                lhs._index_cols,
+                ascending=True,
+                ignore_index=False,
+                na_position="last",
+            )
+
+        lhs = lhs._reset_index_names()
```

Comment:
why? `concat` should to preserve index names when they're not `None`:

<details><summary>Example</summary>

```python
>>> df1
    a
id
0   1
1   2
2   3
3   4
>>> df2
    b
id
0   5
1   6
2   7
3   8
>>> pandas.concat([df1, df2], axis=1)
    a  b
id
0   1  5
1   2  6
2   3  7
3   4  8
```

</details>

Reply:
It seems to preserve it only when joined frames use the same index names. In my tests I used different names and they were removed.

Reply:
Fixed that and added tests.